packages/coreutils/src/schedule.ts

```diff
@@ -0,0 +1,49 @@
+// Copyright (c) Jupyter Development Team.
+// Distributed under the terms of the Modified BSD License.
+
+/**
+ * Test whether the client is a brower and can request animation frames.
+ */
+const BROWSER =
+  typeof document !== 'undefined' &&
+  typeof requestAnimationFrame === 'function';
+
+/**
+ * A composite type for callback handles, useful for unscheduling the callback.
+ */
+export type ScheduleHandle =
+  | ReturnType<typeof requestAnimationFrame>
+  | ReturnType<typeof setImmediate>
+  | ReturnType<typeof setTimeout>;
+
+/**
+ * Schedules a function for invocation as soon as possible asynchronously.
+ *
+ * @param fn The function to invoke when called back.
+ *
+ * @returns A handle that can be used to `unschedule` the `fn` if possible.
+ */
+export function schedule(fn: () => unknown): ScheduleHandle {
+  if (BROWSER && document.visibilityState === 'hidden') {
+    return setTimeout(fn);
+  }
+  if (BROWSER) {
+    return requestAnimationFrame(fn);
+  }
+  return setImmediate(fn);
+}
+
+/**
+ * Unschedules a function invocation if possible.
+ *
+ * @param handle The handle for the callback to unschedule.
+ */
+export function unschedule(handle: ScheduleHandle): void {
+  if (BROWSER && document.visibilityState === 'hidden') {
```

Comment:
I think we need store the original state on the handle, because the document visibility can change between schedule and unschedule.

Reply:
Ugh. I thought about this a lot and I don't want to do it because of the performance hit of tracking the type of function that was invoked. It seems unlikely that in however long it takes to execute an animation frame, the page goes hidden just before the execution completes.

- Keep in mind that this is meant to be near instantaneous and cancel is not guaranteed.
- `unschedule` will not throw an error if it gets an errant handle.

I don't know, what do you think?

Reply:
Yeah that's fair